### PR TITLE
doc: Update 'Secure string handling'

### LINKED ISSUE #20128: Validate user input and keep path in a separate argument for importwallet, createwallet and dumpwallet
**Is your feature request related to a problem? Please describe.**

Wallets with weird names possible which can be exploited in vulnerable web applications that use Bitcoin Core and allow the users to create and import/export wallet

`../testwallet` for Linux
`..\testwallet` for Windows

Tried on Bitcoin Core v 0.20.0

https://github.com/bitcoin/bitcoin/pull/20080#issuecomment-706766527

**Describe the solution you'd like**
+ Keep two arguments: 
`wallet_name` and `wallet_path` for `createwallet` 
`filename` and `filepath` for `dumpwallet` `importwallet`

+ Validate user input. Name should not contain special characters. Path should be optional and if not specified use a default value.

**Describe alternatives you've considered**
Web developers should create secure web apps that use Bitcoin Core.

**Additional context**

image

I understand its not a vulnerability in Bitcoin Core and only affects vulnerable web apps that use Bitcoin Core. However we can at least consider it a bug that may affect something in future or other projects that use Bitcoin Core. Basic checks for user input can improve the security. In 2017 I had a website in which someone had reported a vulnerability that could be used to change price of flight tickets and book with almost zero bitcoin. It was an issue with the website and we had to fix it although nobody could exploit it because the third party APIs that we were using to book the fight tickets were validating all the things. So a ticket couldn't be processed after tampering and changing the price by attacker. Similarly, if any web developer makes a mistake and using Bitcoin Core for the web app would still be unaffected if Bitcoin Core itself doesn't allow such things for wallet names. 

There have been lot of directory traversal related vulnerabilities, recently one was reported in Facebook android app:

https://portswigger.net/daily-swig/vulnerability-in-facebook-android-app-nets-10k-bug-bounty

### PR DESCRIPTION
**Pull request by ghost**
- Add information about possible path traversal attack
- [wallet_name](https://bitcoincore.org/en/doc/0.20.0/rpc/wallet/createwallet/) (string): _The name for the new wallet. If this is a 'path', the wallet will be created at the 'path' location._

Fixes https://github.com/bitcoin/bitcoin/issues/20128 (Not really fixing it but workaround)

This PR is an alternative to https://github.com/bitcoin/bitcoin/pull/20393

### CODE REVIEW

**Review by RiccardoMasutti — APPROVED** [review]
ACK 578315a

**Comment by benthecarman** [comment]
ACK 578315a

**Comment by michaelfolkson** [comment]
I'm not entirely sure what "validating or securely using RPC calls" is referring to. What do you mean by "validating RPC calls" here? And how would one "insecurely" (opposite of securely) use a RPC in this context?

Do you mean it is advised not to reimplement RPCs when building a wallet and to use the ones in Core?

Also a nit: "For example," (what it was originally) is better than "Example"

**Comment by prayank23** [comment]
> I'm not entirely sure what "validating or securely using RPC calls" is referring to. What do you mean by "validating RPC calls" here? 

Validating "string" input as it is mentioned in "secure string handling" section

> And how would one "insecurely" (opposite of securely) use a RPC in this context?

I can share POC if others are okay with it being shared here

> Do you mean it is advised not to reimplement RPCs when building a wallet and to use the ones in Core?

No. I just mean implement them in a way that attacker can't exploit your website by using path traversal.

**Comment by michaelfolkson** [comment]
> Validating "string" input as it is mentioned in "secure string handling" section

So you are not suggesting anything in addition to what is already in the "Secure string handling" section. Validating what is already in the "Secure string handling" section.

> No. I just mean implement them in a way that attacker can't exploit your website by using path traversal.

A suggestion for changes to Bitcoin Core RPCs shouldn't go into docs. They should be discussed in an Issue/PR on their merits which presumably is what is happening in https://github.com/bitcoin/bitcoin/issues/20128. There appears to be confusion though on the exact change(s) you are suggesting there.

> I can share POC if others are okay with it being shared here

I'm not sure this is the best place. Feel free to post on IRC and if it has merits I am sure someone will discuss it with you privately.

**Comment by michaelfolkson** [comment]
As far as I can understand @prayank23 you want to impose additional restrictions on the characters allowed in the wallet name which is being discussed in this PR https://github.com/bitcoin/bitcoin/pull/20393. That is open and currently in draft status. That appears to have some merit and people are engaging with it.

I don't see the motivation for changing the docs at this stage. If a PR based on https://github.com/bitcoin/bitcoin/pull/20393 is merged then maybe additional guidance can be considered for the docs (for whatever restrictions are not merged). But until then this particular doc PR is a Concept NACK from me.

**Comment by prayank23** [comment]
> So you are not suggesting anything in addition to what is already in the "Secure string handling" section. Validating what is already in the "Secure string handling" section.

I have added one example that I found while testing things related to https://github.com/bitcoin/bitcoin/pull/20080#issuecomment-706766527

> A suggestion for changes to Bitcoin Core RPCs shouldn't go into docs

This is not a change to RPC

> They should be discussed in an Issue/PR on their merits which presumably is what is happening in #20128. There appears to be confusion though on the exact change(s) you are suggesting there.

All the confusion and questions have been answered already including the earlier PR. But the solution suggested in earlier is not the best thing to do because it can break few things in lot of projects unnecessarily. And it can be managed better by just mentioning in docs.

> As far as I can understand @prayank23 you want to impose additional restrictions on the characters allowed in the wallet name which is being discussed in this PR #20393. That is open and currently in draft status. That appears to have some merit and people are engaging with it.

Those restrictions was a suggested solution which isn't the best approach to solve this problem IMO. People who are engaging never really tested it so they might find lot of issues with it once they test it on different platforms and projects.

> I don't see the motivation for changing the docs at this stage. If a PR based on #20393 is merged then maybe additional guidance can be considered for the docs (for whatever restrictions are not merged). But until then this particular doc PR is a Concept NACK from me.

Thanks for your review and contribution

**Comment by michaelfolkson** [comment]
So you will be closing https://github.com/bitcoin/bitcoin/pull/20393? You don't think there should be any additional restrictions on the wallet name? Please add a comment to that PR and close it if that is the case.

Instead you want in this doc PR to warn application developers that the Core RPC does not impose these restrictions on the wallet name and hence application developers should be careful to impose application level checks on the wallet names that their users use? Is that correct?

**Comment by prayank23** [comment]
> So you will be closing #20393? You don't think there should be any additional restrictions on the wallet name? Please add a comment to that PR and close it if that is the case.

Yes if this get merged other will be closed. If anyone interested to find a better solution in future can create new PR.

> Instead you want in this doc PR to warn application developers that the Core RPC does not impose these restrictions on the wallet name and hence application developers should be careful to impose application level checks on the wallet names that their users use? Is that correct?

Yes. Correct.

File: doc/JSON-RPC-interface.md [hunk 1]
```diff
@@ -90,11 +90,14 @@ RPC interface will be abused.
   although it does usually provide serialized data using a hex
   representation of the bytes.  If you use RPC data in your programs or
   provide its data to other programs, you must ensure any problem
-  strings are properly escaped.  For example, multiple websites have
+  strings are properly escaped.  Example: multiple websites have
```
[the remark this answers is not in the document]
Comment (prayank23) [reply doc/JSON-RPC-interface.md]: Done. Made changes in https://github.com/bitcoin/bitcoin/pull/20741/commits/cb31b996dd11205942a5fd473c2c1cf0b47a3bf8

**Comment by michaelfolkson** [comment]
Sorry to be a pain but you'll need to [squash](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits) your commits :)

ACK from me otherwise

**Comment by RiccardoMasutti** [comment]
ACK 6e20882

**Review by RiccardoMasutti — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/commit/7117d7503f39f06b74c84777ec4db5d456a8086f

**Comment by prayank23** [comment]
@RiccardoMasutti give me few minutes

**Comment by prayank23** [comment]
@RiccardoMasutti Ignore the other commits and you can ACK this https://github.com/bitcoin/bitcoin/commit/7117d7503f39f06b74c84777ec4db5d456a8086f

**Comment by michaelfolkson** [comment]
ACK 7117d7503f39f06b74c84777ec4db5d456a8086f

Please close https://github.com/bitcoin/bitcoin/pull/20393 and https://github.com/bitcoin/bitcoin/issues/20128 either now or when this is merged if there is nothing further to discuss on them.

**Review by benthecarman — APPROVED** [review]
ACK 7117d7503f39f06b74c84777ec4db5d456a8086f

**Review by luke-jr — APPROVED** [review]
ACK

**Pull request merged**